docs/proposals/resource_management.md

```diff
@@ -0,0 +1,387 @@
+
+
+
+### With Systemd
+
+In systemd we have slices for the three top level QoS class. Further each pod is a subslice of exactly one of the three QoS slices. Each container in a pod belongs to a scope nested under the qosclass-pod slice.
+
+Example:<br>
+We plan to have the following cgroup hierarchy on systemd systems
+
+```
+/memory/G-PodUID.slice/containerUID.scope
+/cpu,cpuacct/G-PodUID.slice/containerUID.scope
+/memory/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/cpu,cpuacct/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/memory/BE.slice/BE-PodUID.slice/containerUID.scope
+/cpu,cpuacct/BE.slice/BE-PodUID.slice/containerUID.scope
+```
+
+There are similar slices for Burstable and Best Effort QoS classes.
+
+## Implementation Plan
+
+The implementation plan is outlined in the next sections.
+We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default atleast in v1.4.
+
+#### Top level Cgroups for QoS tiers
+
+Three top level cgroups for Guaranteed, Burstable and Best effort Qos classes are created when Kubelet starts to run on a node.
+We use libcontainers cgroups manager interface to create/destroy the same. These cgroups are only created once on Kubelet initialization as a part of node setup.
+The Guaranteed pods Cgroup would be nested directly under the root by default.
+But we will have a enable-guaranteed-cgroup flag that would allow for having a specific Guaranteed Cgroup under the root.
+
+#### Pod level Cgroup creation and deletion (Docker runtime)
+
+- When a new pod is brought up, its QoS class is firstly determined.
+- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
+- This interface will be pluggable. Kubelet will support both systemd and raw cgroups based __cgroup__ drivers. We will use the --cgroup-driver flag proposed in the [Systemd Node Spec]().
+- We inject creation and deletion of pod level cgroups into the pod workers.
+- As new pods are added QoS class cgroup parameters are updated to match the resource requests by the Pods.
+- To ensure Kubelet's idempotent behaviour we follow a pod cgroup naming format which is opaque and deterministic. We will use the pod UID as the name of the cgroups.
+
+
+#### Container level cgroups
+
+Have docker manager create container cgroups under pod level cgroups. For determining the parent pod cgroup name and path we have a separate implementation for systemd and raw cgroup systems. This can be easily implemented as now docker has support for specifying parent flag of cgroups. We don't have support for slice management in systemd yet. So we will have to add support for parent flag and general slice management in systemd.
+
+#### Rkt runtime
+
+We want to have rkt create pods under a root QoS class that kubelet specifies, and set pod level cgroup parameters mentioned in this proposal by itself.
```

Comment:
Oops, rkt executes in whatever cgroup it is started in. So, the kubelet + rkt integration just needs to know the cgroup and it will join it. This is easy.